In [1]:
import pandas as pd
dataset = pd.read_csv("StudentsPerformance.csv")
print(dataset)

     gender race/ethnicity      ...      reading score writing score
0    female        group B      ...                 72            74
1    female        group C      ...                 90            88
2    female        group B      ...                 95            93
3      male        group A      ...                 57            44
4      male        group C      ...                 78            75
5    female        group B      ...                 83            78
6    female        group B      ...                 95            92
7      male        group B      ...                 43            39
8      male        group D      ...                 64            67
9    female        group B      ...                 60            50
10     male        group C      ...                 54            52
11     male        group D      ...                 52            43
12   female        group B      ...                 81            73
13     male        group A      ..

In [2]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
gender                         1000 non-null object
race/ethnicity                 1000 non-null object
parental level of education    1000 non-null object
lunch                          1000 non-null object
test preparation course        1000 non-null object
math score                     1000 non-null int64
reading score                  1000 non-null int64
writing score                  1000 non-null int64
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [3]:
dataset["Average"] = (dataset["math score"] + dataset["reading score"] + dataset["writing score"])/3
dataset.head()

gender race/ethnicity    ...     writing score    Average
0  female        group B    ...                74  72.666667
1  female        group C    ...                88  82.333333
2  female        group B    ...                93  92.666667
3    male        group A    ...                44  49.333333
4    male        group C    ...                75  76.333333

[5 rows x 9 columns]

In [4]:
dataset_copy = dataset.copy()
dataset_copy.drop(labels ='parental level of education', axis='columns', inplace=True)
print(dataset_copy.head())

   gender race/ethnicity    ...     writing score    Average
0  female        group B    ...                74  72.666667
1  female        group C    ...                88  82.333333
2  female        group B    ...                93  92.666667
3    male        group A    ...                44  49.333333
4    male        group C    ...                75  76.333333

[5 rows x 8 columns]


In [5]:
dataset_copy.dropna(axis='index', how='any', inplace=True)
print(dataset_copy.head())

   gender race/ethnicity    ...     writing score    Average
0  female        group B    ...                74  72.666667
1  female        group C    ...                88  82.333333
2  female        group B    ...                93  92.666667
3    male        group A    ...                44  49.333333
4    male        group C    ...                75  76.333333

[5 rows x 8 columns]


In [6]:
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(dataset_copy, test_size=0.25, random_state = 123)
print(len(train_set))
print(len(test_set))
print(train_set.head())
print(test_set.head())

750
250
     gender race/ethnicity    ...     writing score    Average
894  female        group E    ...                69  63.333333
941  female        group D    ...                96  88.333333
285    male        group B    ...                82  81.666667
462  female        group E    ...                76  72.333333
370    male        group E    ...                71  77.333333

[5 rows x 8 columns]
     gender race/ethnicity    ...     writing score    Average
131    male        group C    ...                40  43.333333
203  female        group B    ...                68  64.666667
50     male        group E    ...                48  52.000000
585  female        group C    ...                76  72.333333
138  female        group C    ...                67  64.000000

[5 rows x 8 columns]


## Pick an initial set of features for X and the target feature y.  Explain why you made this choice.

I will try to do linear regression for X with Writing score, Math score, Reading score.

And y will be the average of all the three marks


In [7]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
X = train_set[["writing score"]]
Y = train_set["Average"]
reg.fit(X, Y)

print(reg.score(X,Y))

from sklearn.metrics import mean_squared_error
y_pred = reg.predict(X)
mse = mean_squared_error(Y, y_pred)
print("mean squared error is ",  mse)

from math import sqrt
rmse = sqrt(mse)
print("The root mean squared error is ", rmse)

0.9289252096335803
mean squared error is  14.089850381091356
The root mean squared error is  3.7536449460612755


## Notes 
1. Above restults is for X = Writing Score and y is average 
2. Score of this metrics is 0.928 but it should be approx equal to 1, So lets try for remaining attributes.
3. mean squared error is  14.0
4. The root mean squared error is 3.7 

In [8]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
X = train_set[["reading score"]]
Y = train_set["Average"]
reg.fit(X, Y)

print(reg.score(X,Y))

from sklearn.metrics import mean_squared_error
y_pred = reg.predict(X)
mse = mean_squared_error(Y, y_pred)
print("mean squared error is ",  mse)

from math import sqrt
rmse = sqrt(mse)
print("The root mean squared error is ", rmse)

0.9391768932122632
mean squared error is  12.057558945080629
The root mean squared error is  3.4723995946723396


## Notes 
1. Above restults is for X = reading Score and y is average 
2. Score of this metrics is 0.9391 but it should be approx equal to 1, So lets try for remaining attributes.
3. mean squared error is  12.0
4. The root mean squared error is  3.4

In [9]:

from sklearn.linear_model import LinearRegression
reg = LinearRegression()
X = train_set[["math score"]]
Y = train_set["Average"]
reg.fit(X, Y)

print(reg.score(X,Y))

from sklearn.metrics import mean_squared_error
y_pred = reg.predict(X)
mse = mean_squared_error(Y, y_pred)
print("mean squared error is ",  mse)

from math import sqrt
rmse = sqrt(mse)
print("The root mean squared error is ", rmse)

0.8369170659104608
mean squared error is  32.32952400119388
The root mean squared error is  5.685905732703795


## Notes 
1. Above restults is for X = Writing Score and y is average 
2. Score of this metrics is 0.836 but it should be approx equal to 1.
3. mean squared error is  32.3
4. The root mean squared error is 5.6

## Results 

X = reading Score and y = Average is the best metrics from all the above three. as it is almost equal to 1, We can say that is the best.